memtable.cc

```diff
@@ -695,14 +729,12 @@ memtable::make_flat_reader(schema_ptr s,
         }
         auto dk = pos.as_decorated_key();
         auto cr = query::clustering_key_filter_ranges::get_ranges(*s, slice, dk.key());
-        auto snp_schema = snp->schema();
         bool digest_requested = slice.options.contains<query::partition_slice::option::with_digest>();
-        auto rd = make_partition_snapshot_flat_reader<partition_snapshot_read_accounter>(snp_schema, std::move(permit), std::move(dk), std::move(cr), std::move(snp), digest_requested,
-                        *this, _read_section, shared_from_this(), fwd, *this);
+        flat_mutation_reader rd = make_partition_snapshot_flat_reader_from_snp_schema(is_reversed, std::move(permit), std::move(dk), std::move(cr), std::move(snp), digest_requested, *this, _read_section, shared_from_this(), fwd, *this);
         rd.upgrade_schema(s);
```

Comment:
Do we assume that `s` is already reversed for reversed queries?

If yes, then why reverse the schema in the below call to `make_flat_mutation_reader<scanning_reader>`? Is it to obtain the non-reversed schema back for the scanning reader? This doesn't seem right, so I assume that this is not the case (so see below) (but anyway, the end API that we agreed on AFAIK is that the schema passed in to `make_flat_reader` will be reversed for reversed queries).

If not, then shouldn't we reverse the schema for upgrading as well? Because `rd` uses a reversed schema - so when we upgrade the reader, the higher-version schema should also be reversed (so we either upgrade from non-reversed to non-reversed or from reversed to reversed)

Reply:
Actually maybe you do need to unreverse the schema for the scanning reader if you don't support reversed multi-partition queries yet. But if that is the case, it deserves a comment - definitely on the interface of `make_flat_reader` (something like: "if you give a reversed slice but ask for a multi-partition query, you will not get a reversed result, you will get a forward result, and our schema will be unreversed"; see e.g. https://github.com/scylladb/scylla/pull/9281/commits/9931251ac668b0dc733b3f57437cd5b0e9c967ef#diff-2bea054e8aec03350976350cbde49af18fdba097007f0cb27dd9b2c0a4d37756R30 for a similarly weird API), and a comment on the `make_reversed()` call below that we're actually unreversing the schema.